### PR TITLE
fix for #11424

### LINKED ISSUE #11424: script compiler options can cause warnings when starting REPL
## Compiler version
        Scala compiler version 3.0.0-M4-bin-SNAPSHOT-git-668e9a5 -- Copyright 2002-2021, LAMP/EPFL

## Minimized code

The `-save` option should not also pass to the REPL:
```Scala
SCALA_OPTS="-save" scala
```

## Output

```scala
$ SCALA_OPTS="-save" bin/scala
Starting scala3 REPL...
bad option '-save' was ignored
```

## Expectation
`-save` and `-compile-only` options should only be seen by scripting

### PR DESCRIPTION
**Pull request by philwalk**
this fixes #11424 
manually verified for currently defined dotc script-only options:
```
-save
-compile-only
```

### CODE REVIEW

**Comment by liufengyun** [comment]
I think we need to have this in RC1, given that it is user-facing for anyone trying the repl.

**Comment by michelou** [comment]
> I think we need to have this in RC1, given that it is user-facing for anyone trying the repl.

Note also : 
- line 126 : `java`
- line 151 : `$JAVACMD`

Question : Did anybody test script file [`scala`](https://github.com/lampepfl/dotty/blob/d166b568e2c5cdf1522df21a519a8a666692487c/dist/bin/scala)
 on Mingw64 (Windows), Cygwin (Windows) or any other flavor of Linux ?!

**Comment by liufengyun** [comment]
Good catch @michelou . We have a test for distribution https://github.com/lampepfl/packtest. I think it's time to have that test as part of CI.

**Review by anatoliykmetyuk — APPROVED** [review]

**Comment by philwalk** [comment]
@michelou - I tested in these environments:
```
Linux x 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
Linux x 5.4.0-62-generic #70~18.04.1-Ubuntu SMP Tue Jan 12 17:18:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
CYGWIN_NT-10.0 x 3.1.7(0.340/5/3) 2020-08-22 17:48 x86_64 Cygwin
```
Although my `msys` environment is possibly misconfigured, it's broken independent of this PR.
The REPL doesn't appear to be working in general on msys/mingw.
```
$ uname -a
MINGW64_NT-10.0-19042 x 3.1.7-340.x86_64 2020-10-23 13:08 UTC x86_64 Msys
$ scala
Starting scala3 REPL...
Exception in thread "main" java.lang.IllegalStateException: Unable to create a system terminal
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:317)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:263)
        at dotty.tools.repl.JLineTerminal.<init>(JLineTerminal.scala:23)
        at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:108)
        at dotty.tools.repl.Main$.main(Main.scala:6)
        at dotty.tools.repl.Main.main(Main.scala)
        Suppressed: com.sun.jna.LastErrorException: [6] The handle is invalid.
                at com.sun.jna.Native.invokeVoid(Native Method)
                at com.sun.jna.Function.invoke(Function.java:415)
                at com.sun.jna.Function.invoke(Function.java:361)
                at com.sun.jna.Library$Handler.invoke(Library.java:265)
                at org.jline.terminal.impl.jna.win.$Proxy0.GetConsoleMode(Unknown Source)
                at org.jline.terminal.impl.jna.win.JnaWinSysTerminal.createTerminal(JnaWinSysTerminal.java:40)
                at org.jline.terminal.impl.jna.JnaSupportImpl.winSysTerminal(JnaSupportImpl.java:39)
                at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:326)
                ... 5 more
        Suppressed: java.util.NoSuchElementException
                at java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1308)
                at java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1296)
                at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1394)
                at org.jline.terminal.TerminalBuilder.load(TerminalBuilder.java:487)
                at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:334)
                ... 5 more
```

**Comment by michelou** [comment]
> @michelou - I tested in these environments:
> 
> ```
> Linux x 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
> Linux x 5.4.0-62-generic #70~18.04.1-Ubuntu SMP Tue Jan 12 17:18:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
> CYGWIN_NT-10.0 x 3.1.7(0.340/5/3) 2020-08-22 17:48 x86_64 Cygwin
> ```
> 
> Although my `msys` environment is possibly misconfigured, it's broken independent of this PR.
> The REPL doesn't appear to be working in general on msys/mingw.
> 
> ```
> $ uname -a
> MINGW64_NT-10.0-19042 x 3.1.7-340.x86_64 2020-10-23 13:08 UTC x86_64 Msys
> $ scala
> Starting scala3 REPL...
> Exception in thread "main" java.lang.IllegalStateException: Unable to create a system terminal
>         at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:317)
>         at (TerminalBuilder.java:263)
>         [...]
>                 ... 5 more
> ```

@philwalk I know the issue (I was waiting for someone else to also report it). In my case I made the following local change on line 119 of file [`dist/bin/common`](https://github.com/lampepfl/dotty/blob/master/dist/bin/common) : 

` [[ $mingw ]] || JNA=$(find_lib "*jna-5*")`

instead of

`JNA=$(find_lib "*jna-5*")`

**Comment by philwalk** [comment]
@michelou - I can do a PR for the `$mingw` fix if it's not already in the works

**Comment by michelou** [comment]
> @michelou - I can do a PR for the `$mingw` fix if it's not already in the works

I guess my fix works for you as well.
Then you can do a PR (I'm myself focusing on [`scala.bat`](https://github.com/michelou/dotty-examples/tree/master/bin/3.0.0)).

**Comment by philwalk** [comment]
@michelou the fix worked for MINGW but not for MSYS2, so I haven't done the PR yet.   The problem is similar but not identical.
What led you to the proposed fix?
```
$ uname -a
MSYS_NT-10.0-19042 d5 3.1.7-340.x86_64 2020-11-08 12:32 UTC x86_64 Msys
philwalk@d5:~/workspace/dotty
$ scala
Starting scala3 REPL...
Exception in thread "main" java.lang.IllegalStateException: Unable to create a system terminal
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:317)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:263)
        at dotty.tools.repl.JLineTerminal.<init>(JLineTerminal.scala:23)
        at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:108)
        at dotty.tools.repl.Main$.main(Main.scala:6)
        at dotty.tools.repl.Main.main(Main.scala)
        Suppressed: java.lang.NoClassDefFoundError: com/sun/jna/LastErrorException
                at org.jline.terminal.impl.jna.JnaSupportImpl.isWindowsConsole(JnaSupportImpl.java:44)
                at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:325)
                ... 5 more
        Caused by: java.lang.ClassNotFoundException: com.sun.jna.LastErrorException
                at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
                at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
                at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
                ... 7 more
        Suppressed: java.util.NoSuchElementException
                at java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1308)
                at java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1296)
                at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1394)
                at org.jline.terminal.TerminalBuilder.load(TerminalBuilder.java:487)
                at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:334)
                ... 5 more

**Comment by michelou** [comment]
@philwalk Here is a first observation : the shell script [`dist/bin/common`](https://github.com/lampepfl/dotty/blob/master/dist/bin/common) currently handles the 3 cases `cygwin`, `mingw` and `darwin`. Did you consider adding a case `msys` ?

In my case I get `MINGW...` when running `uname` with my MSYS2 (64-bit) installation. I need to search how one can get `MSYS...` when running `uname`. Can you provide me the download link of the software installer you used to create your MSYS environment ?

**Comment by philwalk** [comment]
@michelou
I get `mingw` here: [https://gitforwindows.org/](https://gitforwindows.org/), also called `git-bash` by `conemu`.
I get `msys2` here: [https://www.msys2.org/](https://www.msys2.org/)
Here's  a comparison of `msys2` with `git-for-windows`:
[https://github.com/git-for-windows/build-extra#the-difference-between-msys2-and-mingw](https://github.com/git-for-windows/build-extra#the-difference-between-msys2-and-mingw) 

Both `msys2` and `mingw` report the same thing (`MSys`) in response to `uname -o`
On my system, all three `cygwin`, `mingw` and `msys2` provide `cygpath`, as shown here:
```scala
$ uname
MINGW64_NT-10.0-19042
$ cygpath -m /
C:/opt/gitbash/

$ uname
MSYS_NT-10.0-19042
$ cygpath -m /
C:/msys64/

$ uname
CYGWIN_NT-10.0
$ cygpath -m /
C:/cygwin64
```
So a simplification to [dist/bin/common](https://github.com/lampepfl/dotty/blob/master/dist/bin/common) would be:
```scala
# For Cygwin, Mingw and MSys, switch paths to Windows-mixed format before running java
if [[ $cygwin || $mingw || $msys ]]; then
  [ -n "$PROG_HOME" ] &&
    PROG_HOME=`cygpath -am "$PROG_HOME"`
  [ -n "$JAVA_HOME" ] &&
    JAVA_HOME=`cygpath -am "$JAVA_HOME"`
  CLASSPATH_SUFFIX=";"
  PSEP=";"
fi
```

**Comment by michelou** [comment]
@philwalk Looking at the Scala 2 Unix template ([`tool-unix.tmpl`](https://github.com/scala/scala/blob/2.13.x/src/compiler/templates/tool-unix.tmpl)) used for generating the bash scripts `scala` and `scalac` is also worth :
- it defines/uses the three variables `cygwin`, `mingw` and `msys` (see [line 64 ff](https://github.com/scala/scala/blob/2.13.x/src/compiler/templates/tool-unix.tmpl#L64))
- it performs some action in relation with the usage of Jline terminal (see [line 109](https://github.com/scala/scala/blob/eb8cb18cd7b1cae3629448ef512558d8e2496f4d/src/compiler/templates/tool-unix.tmpl#L109))

IMO introducing variable `msys` is a good idea.
But two questions now have to be examined : 
- why is the `mingw` case (line 73) handled differently from `cygwin` (line 83) in file [`dist/bin/common`](https://github.com/lampepfl/dotty/blob/master/dist/bin/common) ?
- would the option `WINDOWS_OPT="-Djline.terminal=unix` (line 113) solve our concern ?

In conclusion we need to investigate further.

**Comment by philwalk** [comment]
@michelou 
I incorporated the scala2 approach and it tests well.

> why is currently the mingw case (line 73) handled differently from cygwin (line 83) in file dist/bin/common ?

The differences I have discovered between `cygwin` and either `mingw` or `msys` are these:
1. `cygwin` always provides the `cygpath` executable for converting between posix-like and `windows` paths
2. early versions (and possibly some current versions?) of `mingw` and `msys` did not provide `cygpath`, although many now do

The problem that [dist/bin/common](https://github.com/lampepfl/dotty/blob/master/dist/bin/common) (line 73) addresses is better handled by `cygpath` if it's available.   The difference is that (line 73) only converts between `\` and `/` but the `/etc/fstab` file can also cause require directory mounts to be translated.   The `cygpath` tool handles both the `/` path separator and `/etc/fstab` mounts.

Here's an approach that seems likely to provide a better result.
```bash
$ has_cygpath=false
$ if `which cygpath.exe > /dev/null 2>&1`; then has_cygpath=true ; fi
```
Then, instead of  [dist/bin/common](https://github.com/lampepfl/dotty/blob/master/dist/bin/common) (lines 73-105):
```bash
# For Cygwin, switch paths to Windows-mixed format before running java
if $has_cygpath; then
  [ -n "$PROG_HOME" ] &&
    PROG_HOME=`cygpath -am "$PROG_HOME"`
  [ -n "$JAVA_HOME" ] &&
    JAVA_HOME=`cygpath -am "$JAVA_HOME"`
  CLASSPATH_SUFFIX=";"
  PSEP=";"
elif ($mingw || $msys); then
  # For Mingw / Msys, ensure paths are in UNIX format before anything is touched
  [ -n "$PROG_HOME" ] &&
    PROG_HOME="`(cd "$PROG_HOME"; pwd -W | sed 's|/|\\\\|g')`"
  [ -n "$JAVA_HOME" ] &&
    JAVA_HOME="`(cd "$JAVA_HOME"; pwd -W | sed 's|/|\\\\|g')`"
  CLASSPATH_SUFFIX=";"
  PSEP=";"
fi

#/*--------------------------------------------------
# * The code below is for Dotty
# *-------------------------------------------------*/

find_lib () {
  local lib=$(find $PROG_HOME/lib/ -name "$1")
  if $has_cygpath; then
    cygpath -am $lib
  elif $mingw; then
    echo $lib | sed 's|/|\\\\|g'
  else
    echo $lib
  fi
}
```
